### DISCUSSION TITLE
Help Using react-query

### DISCUSSION THREAD

**Discussion by evmartinelli** [TanStack/query#3404] [Q&A] [answered]
I updated my react packages and now im getting error when I try to use this code:

Rest Call
```
import { CategoriesQueryOptionsType, Category } from "@framework/types";
import http from "@framework/utils/http";
import { API_ENDPOINTS } from "@framework/utils/api-endpoints";
import { useQuery } from "react-query";

export const fetchCategories = async ({ queryKey }: any) => {
	const [_key, _params] = queryKey;
	const {
		data: { data },
	} = await http.get(API_ENDPOINTS.CATEGORIES);
	return { categories: { data: data as Category[] } };
};
export const useCategoriesQuery = (options: CategoriesQueryOptionsType) => {
	return useQuery<{ categories: { data: Category[] } }, Error>(
		[API_ENDPOINTS.CATEGORIES, options],
		fetchCategories
	);
};
```

Usage
```
	const { data, isLoading, error } = useCategoriesQuery({
		limit: 10,
	});
```

`{isLoading && !data?.categories?.data?.length` -> error: Type error: Property 'categories' does not exist on type 'never'.

If I omit the "isLoading"" the error dissapear.

Can anyone help me, please?

**Comment by TkDodo** [comment]
Did you update to typescript 4.6 ?

**Comment by TkDodo** [comment, marked as answer]
I've put your code into typescript playground:

- [no error on TS 4.5](https://www.typescriptlang.org/play?ts=4.5.5#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1QJ5wC+cAZlBCHAERToCGAxjALQCOO+VA3AFA-oAPSLDgMIAO1TwS6GAwAWAYTox0Ac2jB0qOAF44dVHnEM4ACiTtceANLoChAFwHxeAJR6AfIh4BIMZLwANoA+gDW9gA0cCFgdFB0IKgAunpwVvh2eLz+ElI+vr4AJip0kX7E+nQA7nTA8AAK5CDAGAB0tKgQADYAbugWcCUwdM5BqYRufrQwyFDiiKIq6prazkjDdEREvIS8gsLwAflo6MqqGlBaqNjWaWYeut4I07JzC6e3+AA8SAzLl2u6yGpWcAEEoAk8N9kOIwuIINVxN5iIRogBRSHQTxmPy+IIAchIEAgBOS5UKMjkSgBq1Qfjcu14fmO8A2pWirQAMhA6EVgOI1NFcOQoNt9KdzisrtovngHsy4HBgCRzNzefzBXAAGTauAAQk2AH42v8LnSTca2t10IKYPIPC8lUqeErCEA)
- [an error on TS 4.6](https://www.typescriptlang.org/play?ts=4.6.2#code/JYWwDg9gTgLgBAbzgVwM4FMCKz1QJ5wC+cAZlBCHAERToCGAxjALQCOO+VA3AFA-oAPSLDgMIAO1TwS6GAwAWAYTox0Ac2jB0qOAF44dVHnEM4ACiTtceANLoChAFwHxeAJR6AfIh4BIMZLwANoA+gDW9gA0cCFgdFB0IKgAunpwVvh2eLz+ElI+vr4AJip0kX7E+nQA7nTA8AAK5CDAGAB0tKgQADYAbugWcCUwdM5BqYRufrQwyFDiiKIq6prazkjDdEREvIS8gsLwAflo6MqqGlBaqNjWaWYeut4I07JzC6e3+AA8SAzLl2u6yGpWcAEEoAk8N9kOIwuIINVxN5iIRogBRSHQTxmPy+IIAchIEAgBOS5UKMjkSgBq1Qfjcu14fmO8A2pWirQAMhA6EVgOI1NFcOQoNt9KdzisrtovngHsy4HBgCRzNzefzBXAAGTauAAQk2AH42v8LnSTca2t10IKYPIPC8lUqeErCEA)

TS 4.6 has a new feature for better [Control Flow Analysis for Destructured Discriminated Unions](https://devblogs.microsoft.com/typescript/announcing-typescript-4-6/#cfa-destructured-discriminated-unions), which is exactly what's kicking in here. Since you are checking for `isLoading`, and react-query's types are defined as `data` being `undefined` if you are in `loading` state, the compiler now knows that the length check is unnecessary, because `data` cannot be an array during loading state :)

**Reply by evmartinelli** [reply]
thanks!!!
